main.js

```diff
@@ -1048,21 +1061,58 @@ Creep.prototype.travelToRoom = function travelToRoom(tgtRoom, forward) {
 		return ERR_NO_PATH;
 
 	let list_route = _.get(this, ["memory", "list_route"]);
-	if (list_route != null) {
-		if (forward == true) {
+	
+	if (list_route && list_route.length > 0) {
+		if (forward === true) {
 			for (let i = 1; i < list_route.length; i++) {
-				if (_.get(list_route, i) != null && this.room.name == _.get(list_route, i - 1)) {
+				if (this.room.name === list_route[i - 1]) {
+					// Check for portal to next room in the route
+					let portal = this.pos.findClosestByRange(FIND_STRUCTURES, {
+						filter: (structure) => structure.structureType == STRUCTURE_PORTAL && structure.destination.roomName == list_route[i]
+					});
+					if (portal) {
+						_.set(this, ["memory", "path", "portal"], portal.id);
+						let result = this.travel(portal.pos);
+						if (result == OK) {
+							return OK;
+						} else {
+							// Fallback to moveTo if travel fails
+							result = this.moveTo(portal.pos);
+							if (result == OK) {
+								return OK;
+							}
+						}
+					}
+					// Attempt to travel to the exit tile of the target room
 					let result = this.travelToExitTile(list_route[i]);
 					if (result == OK)
 						return OK;
 					else
 						return this.travel(new RoomPosition(25, 25, list_route[i]));
 				}
 			}
-		} else if (forward == false) {
+		} else if (forward === false) {
 			for (let i = list_route.length - 2; i >= 0; i--) {
-				if (_.get(list_route, i) != null && this.room.name == _.get(list_route, i + 1)) {
-					let result = this.travelToExitTile(list_route[i])
+				if (this.room.name === list_route[i + 1]) {
+					// Check for portal to previous room in the route
+					let portal = this.pos.findClosestByRange(FIND_STRUCTURES, {
```

Comment:
Note for future reference- this is probably a CPU-expensive routine to run every tick. Can be optimized when needed with caching.